### PR TITLE
TRUNK-5367: Create URL global property on OpenMRS core

### PR DESCRIPTION
**Pull request by mwaz**
## Description of what I changed
I added a global property LOGIN URL to hold the default login URL and be accessible publicly.

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-5367

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 1]
```diff
@@ -1081,6 +1081,8 @@ public static String getOpenmrsProperty(String property) {
 		
 		props
 				.add(new GlobalProperty(GP_DRUG_ORDER_DRUG_OTHER, "", "Specifies the uuid of the concept which represents drug other non coded"));
+		props.add(new GlobalProperty(LOGIN_URL, "../../login.htm",
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: Why not just login.html?

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 2]
```diff
@@ -1235,6 +1237,8 @@ public static String getOpenmrsProperty(String property) {
 	
 	public static final String LOG_LEVEL_FATAL = "fatal";
 	
+	public static final String LOGIN_URL = "../../login.htm";
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: Why not just login.html?

**Comment by mwaz** [comment]
@dkayiwa resolved.

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 3]
```diff
@@ -1081,6 +1081,8 @@ public static String getOpenmrsProperty(String property) {
 		
 		props
 				.add(new GlobalProperty(GP_DRUG_ORDER_DRUG_OTHER, "", "Specifies the uuid of the concept which represents drug other non coded"));
+		props.add(new GlobalProperty(LOGIN_URL, "../../login.html",
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: What is the use of ../../

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 4]
```diff
@@ -1235,6 +1237,8 @@ public static String getOpenmrsProperty(String property) {
 	
 	public static final String LOG_LEVEL_FATAL = "fatal";
 	
+	public static final String LOGIN_URL = "../../login.html";
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: What is the use of ../../

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java (hunk 3 above)
Comment (dkayiwa) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: The default value should be login.htm

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 5]
```diff
@@ -1081,6 +1081,8 @@ public static String getOpenmrsProperty(String property) {
 		
 		props
 				.add(new GlobalProperty(GP_DRUG_ORDER_DRUG_OTHER, "", "Specifies the uuid of the concept which represents drug other non coded"));
+		props.add(new GlobalProperty(LOGIN_URL, "login.html",
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: The default should be login.htm

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 6]
```diff
@@ -1081,6 +1081,8 @@ public static String getOpenmrsProperty(String property) {
 		
 		props
 				.add(new GlobalProperty(GP_DRUG_ORDER_DRUG_OTHER, "", "Specifies the uuid of the concept which represents drug other non coded"));
+		props.add(new GlobalProperty(LOGIN_URL, "login.htm",
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: Do you need to write login.htm in two places? 😊

Comment (mwaz) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: Nope, one of them is just a description of the key pair value of the global properties.

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 7]
```diff
@@ -1081,6 +1081,8 @@ public static String getOpenmrsProperty(String property) {
 		
 		props
 				.add(new GlobalProperty(GP_DRUG_ORDER_DRUG_OTHER, "", "Specifies the uuid of the concept which represents drug other non coded"));
+		props.add(new GlobalProperty(LOGIN_URL, "",
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: My comment did not mean removing the default value. Am just guiding you to something else. 😊

Comment (mwaz) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: kindly elaborate..

Comment (mwaz) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: do you mean removing the global declaration and combining it with the addition on props.add(new GlobalProperty (...... property here))

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: First of all, get it back to what it was. Then show my original comment to your fellow apprentices and ask them what i could be driving at. 😊

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 8]
```diff
@@ -1081,6 +1081,8 @@ public static String getOpenmrsProperty(String property) {
 		
 		props
 				.add(new GlobalProperty(GP_DRUG_ORDER_DRUG_OTHER, "", "Specifies the uuid of the concept which represents drug other non coded"));
+		props.add(new GlobalProperty(LOGIN_URL, LOGIN_URL,
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: Did you take a look at the expected parameters for the GlobalProperty constructor?

Comment (mwaz) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: yea it expects a property, a value, and a description

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: Can you tell me the actual value that you are passing in for each of those?

Comment (mwaz) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: the value is the login.htm which is declared as a public by the LOGIN_URL

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: I mean value for each of the parameters.
1) property, 2) value, 3) description

Comment (mwaz) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: 1. login.htm, 2) login.htm, 3) "Responsible for defining the Authentication URL."

Mhh something seems wrong on the property and the value.

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: That is exactly what i wanted you to see for yourself. 😊

Comment (mwaz) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: thanks.

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 9]
```diff
@@ -1235,6 +1237,10 @@ public static String getOpenmrsProperty(String property) {
 	
 	public static final String LOG_LEVEL_FATAL = "fatal";
 	
+	public static final String LOGIN_URL = "login.htm";
+	
+	public static final String LOGIN_URL_PROPERTY = "LOGIN_URL";
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: Did you take a look at this class to see the convention used for naming other global property constants?

Comment (mwaz) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: had not, now i have

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 10]
```diff
@@ -1235,6 +1237,10 @@ public static String getOpenmrsProperty(String property) {
 	
 	public static final String LOG_LEVEL_FATAL = "fatal";
 	
+	public static final String LOGIN_URL = "login.htm";
+	
+	public static final String GP_LOGIN_URL = "LOGIN_URL";
```
[the remark this answers is not in the document]
Comment (mwaz) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: there are two types of global property naming, how does one differentiate between using the GP or GLOBAL_PROPERTY prefix or just using the name such as "allergy.concept.reactions"

[the remark this answers is not in the document]
Comment (dkayiwa) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: Does that page have anything like GO or GLOBAL_PROPERTY prefix? 😊

[the remark this answers is not in the document]
Comment (mwaz) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: the above worked.

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 11]
```diff
@@ -1235,6 +1237,10 @@ public static String getOpenmrsProperty(String property) {
 	
 	public static final String LOG_LEVEL_FATAL = "fatal";
 	
+	public static final String LOGIN_URL = "login.htm";
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: Can you add some javadoc to these two constants?

Comment (mwaz) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: sure

Comment (mwaz) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: LOGIN_URL AND GP_LOGIN_URL

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 12]
```diff
@@ -1234,6 +1236,23 @@ public static String getOpenmrsProperty(String property) {
...
+	 */
+	public static final String LOGIN_URL = "login.htm";
+	
+	/**
+	 * Global property name that defines the default url 
+	 * responsible for authentication if user is not logged in.
+	 *
+	 *  @see #LOGIN_URL
+	 */
+
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: Can you remove the above empty line?

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 13]
```diff
@@ -1234,6 +1236,23 @@ public static String getOpenmrsProperty(String property) {
 	public static final String LOG_LEVEL_ERROR = "error";
 	
 	public static final String LOG_LEVEL_FATAL = "fatal";
+
+	/**
+	 * Global property string value which is the file name provided 
+	 * after successful routing to the default authentication page.
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: Is the above description correct?

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 14]
```diff
@@ -1234,6 +1236,22 @@ public static String getOpenmrsProperty(String property) {
 	public static final String LOG_LEVEL_ERROR = "error";
 	
 	public static final String LOG_LEVEL_FATAL = "fatal";
+
+	/**
+	 * Global property value which defines the default file
```
[the remark this answers is not in the document]
Comment (mwaz) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: closer home?

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 15]
```diff
@@ -1234,6 +1236,22 @@ public static String getOpenmrsProperty(String property) {
 	public static final String LOG_LEVEL_ERROR = "error";
 	
 	public static final String LOG_LEVEL_FATAL = "fatal";
+
+	/**
+	 * Global property value which defines the default authentication
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: GP_LOGIN_URL is well described.
But the description for LOGIN_URL is still not matching what GP_LOGIN_URL talks about.
Can you take a look an existing properties descriptions for an example?
Feel free to ask your team mates for help.

Comment (mwaz) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: resolved?

**Pull request merged**